src/playermat/Playmat.ttslua

```diff
@@ -93,293 +70,271 @@ function onload(save_state)
-
-    getDrawDiscardDecks()
-
-    if investigator == "Norman Withers" then
-        local harbinger = false
-        if topCard ~= nil and topCard.getName() == "The Harbinger" then
-            harbinger = true
-        else
-            if drawDeck ~= nil and not drawDeck.is_face_down then
-                local cards = drawDeck.getObjects()
-                local bottomCard = cards[#cards]
-                if bottomCard.name == "The Harbinger" then
-                    harbinger = true
-                end
-            end
-        end
+---------------------------------------------------------
+-- Upkeep button
+---------------------------------------------------------
 
-        if harbinger then
-            printToColor("The Harbinger is on top of your deck, not drawing cards", messageColor)
-            return -1
-        end
+function doUpkeep(_, color, alt_click)
+  setMessageColor(color)
 
-        if topCard ~= nil then
-            topCard.deal(numCards, PLAYER_COLOR)
-            numCards = numCards - 1
-            if numCards == 0 then return end
+  -- right-click binds to new player color
+  if alt_click then
+    PLAYER_COLOR = color
+    printToColor("Upkeep button bound to " .. color, color)
+    return
+  end
+  
+  local forcedLearning = false
+
+  -- unexhaust cards in play zone, flip action tokens and find forcedLearning
+  for _, v in ipairs(searchArea(PLAY_ZONE_POSITION, PLAY_ZONE_SCALE)) do
+    local obj = v.hit_object
+    if obj.tag == "Card" and not obj.is_face_down and not doNotReady(obj) then
+      local notes = JSON.decode(obj.getGMNotes()) or {}
+      if notes.id == "08031" then
+        forcedLearning = true
+      elseif notes.type ~= "Investigator" then
+        obj.setRotation(PLAY_ZONE_ROTATION)
+
+        -- check for cards with 'replenish' in their metadata
+        if notes.uses ~= nil then
+          local count = notes.uses[1].count
+          local replenish = notes.uses[1].replenish
+          if count and replenish then replenishTokens(obj, count, replenish) end
         end
+      end
+    elseif obj.getDescription() == "Action Token" and obj.is_face_down then
+      obj.flip()
     end
+  end
 
-    local deckSize
-    if drawDeck == nil then
-        deckSize = 0
-    elseif drawDeck.tag == "Deck" then
-        deckSize = #drawDeck.getObjects()
-    else
-        deckSize = 1
-    end
-
-    if deckSize >= numCards then
-        drawCards(numCards)
-        return
+  -- flip investigator mini-card if found
+  -- flip summoned servitor mini-cards (To-Do: don't flip all of them)
+  if activeInvestigatorId ~= nil then
+    local miniId = string.match(activeInvestigatorId, "%d%d%d%d%d") .. "-m"
+    for _, obj in ipairs(getObjects()) do
+      if obj.tag == "Card" then
+        local notes = JSON.decode(obj.getGMNotes())
+        if obj.is_face_down and notes ~= nil and notes.type == "Minicard" and (notes.id == miniId or notes.id == "09080-m") then
```

Comment:
Can move the obj.is_face_down to the same check with tag == Card, will avoid the JSON decode time if the object wouldn't flip anyway

Reply:
Good catch, will change later